### PR TITLE
[Feat] Add checksum for donwload and upload file methods

### PR DESCRIPTION
**Pull request by alejoxbg**
This PR adds a function that checks if the binaries of the files to be uploaded or downloaded are the same, if they are, it skips downloading/uploading them, when it finds differences or does not find the file it downloads/uploads them. To test it I use the library with a simple test file, download and upload the file and run it several times to verify that it was not uploading when it did not change and that when it did not exist or when there was a change it performed the appropriate action.

### CODE REVIEW

File: gcp_io/gcp_interface.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import base64
+import hashlib
 import binascii
 import datetime
 import os
+from pickle import TRUE
```
Comment (charlielito) [inline gcp_io/gcp_interface.py]: Remove this since this is not used

File: gcp_io/gcp_interface.py [hunk 2]
```diff
@@ -151,6 +153,27 @@ def get_md5sum(self, file: str) -> str:
...
+    def check_sum(self, gcs_file: str, local_file: str) -> bool:
+        
+        """! Check if there are differences between a local file and one in GCP
+        @param gcs_file (str) Full path to the file in cloud storage.
+        @param local_file (str) Full path to the local file.
+        """
+        # If file exists, check if it has different hash
+        if os.path.exists(local_file):
+            file_content = self.get_blob(gcs_file).download_as_string()
+            remote_filehash = hashlib.sha256(file_content).hexdigest()
```
Comment (charlielito) [inline gcp_io/gcp_interface.py]: This does not make much sense because you are downloading the flle anyway. See how we are doing it in Kronos https://github.com/kiwicampus/Kronos-Project/blob/develop/rover/ros2/src/python_utils/python_utils/io.py#L326

File: gcp_io/gcp_interface.py [hunk 3]
```diff
@@ -199,21 +222,35 @@ def download_file(
         self,
         src_file: str,
         dst_file: str,
+        checks_cahnges: bool = True,
```
Comment (charlielito) [inline gcp_io/gcp_interface.py]: Typo

File: gcp_io/gcp_interface.py [hunk 4]
```diff
@@ -199,21 +222,35 @@ def download_file(
...
     ):
         """! Downloads the file from cloud storage to local file
         @param src_file (str) Full path to the file in cloud storage.
         @param dst_file (str) Full path to the file to be downloaded.
+        @param checks_changes (bool) Option to check if the file has changed.
         """
+        if checks_cahnges:
```
Comment (charlielito) [inline gcp_io/gcp_interface.py]: Typo

**Review by charlielito — CHANGES_REQUESTED** [review]

File: gcp_io/gcp_interface.py (hunk 2 above)
Comment (charlielito) [reply gcp_io/gcp_interface.py]:
```python
def get_md5sum(file: str) -> str:
    """! Gets md5sum of file
    @param file (str) Full path to file
    @returns (str) md5sum
    """
    if "gs://" in file:
        md5_hash = binascii.hexlify(
            base64.urlsafe_b64decode(get_gcp_blob(file).md5_hash)
        ).decode("utf-8")
    else:
        md5_hash = md5sum(file)
    return md5_hash
  ````

File: gcp_io/gcp_interface.py [hunk 5]
```diff
@@ -1,9 +1,8 @@
+from ast import Return
```
Comment (charlielito) [inline gcp_io/gcp_interface.py]: what is this??

File: test/test_files.py [hunk 6]
```diff
@@ -1,8 +1,10 @@
 import unittest
 import urllib.request
+import os
+import json
 from typing import Any, Dict
 
-from gcp_io import GCPInterface
+from gcp_io.gcp_interface import GCPInterface
```
Comment (charlielito) [inline test/test_files.py]: why did you change this?

File: gcp_io/gcp_interface.py [hunk 7]
```diff
@@ -222,36 +214,35 @@ def download_file(
...
         """! Downloads the file from cloud storage to local file
         @param src_file (str) Full path to the file in cloud storage.
         @param dst_file (str) Full path to the file to be downloaded.
         @param checks_changes (bool) Option to check if the file has changed.
         """
-        if checks_cahnges:
-            if self.check_sum(src_file, dst_file):
-                return
+        if checks_changes and self.check_sum(src_file, dst_file):
+            return False
```
Comment (charlielito) [inline gcp_io/gcp_interface.py]: Don't return a bool value. The function spec doesn't return anything

File: gcp_io/gcp_interface.py [hunk 8]
```diff
@@ -222,36 +214,35 @@ def download_file(
...
         #print("Downloading file")
         content_bytes = self.get_bytes(src_file)
         with open(dst_file, "wb") as f:
             f.write(content_bytes)
+        return True
```
Comment (charlielito) [inline gcp_io/gcp_interface.py]: Same here

File: gcp_io/gcp_interface.py [hunk 9]
```diff
@@ -222,36 +214,35 @@ def download_file(
...
         @param checks_changes (bool) Option to check if the file has changed.
         """
-        if checks_cahnges:
-            if self.check_sum(gcs_file, local_file):
-                return
-        #print("Uploading file")
+        if checks_changes and self.check_sum(gcs_file, local_file):
+            return False
         self.blob(gcs_file).upload_from_filename(local_file)
+        return True
```
Comment (charlielito) [inline gcp_io/gcp_interface.py]: Same for this function, don't return a boolean

File: test/test_files.py [hunk 10]
```diff
@@ -20,3 +22,37 @@ def test_signed_url_error(self):
         with self.assertRaises(Exception):
             signed_url = interface.gcs2signed(gcs_url, expiration_mins=1)
             assert signed_url is None
+    
+    def test_dowload_files(self):
+        gcs_url = "gs://gcp-io-tests/files/test_donwload.json"
+        downloaded_file = interface.download_file(gcs_url, "test_donwload.json")
+        os.remove("test_donwload.json") 
+        assert downloaded_file is True
```
Comment (charlielito) [inline test/test_files.py]:
```suggestion
        assert downloaded_file
```

File: test/test_files.py [hunk 11]
```diff
@@ -20,3 +22,37 @@ def test_signed_url_error(self):
...
+
+    def test_repeated_dowload_files(self):
+        gcs_url = "gs://gcp-io-tests/files/test_donwload.json"
+        interface.download_file(gcs_url, "test_donwload.json")
+        downloaded_file = interface.download_file(gcs_url, "test_donwload.json")
+        os.remove("test_donwload.json") 
+        assert downloaded_file is False
```
Comment (charlielito) [inline test/test_files.py]:
```suggestion
        assert not downloaded_file
```

Comment (charlielito) [reply test/test_files.py]: Same for the rest

**Review by charlielito — CHANGES_REQUESTED** [review]
I'll prefer to change the name from `checks_changes` to something like `md5sum_check`

Also because you shouldn't return a boolean in `upload_file` and `download_file` functions you need to refactor the tests :(

File: gcp_io/gcp_interface.py (hunk 5 above)
Comment (alejoxbg) [reply gcp_io/gcp_interface.py]: maybe some automatic vscode import, I will remove it.

**Review by charlielito — APPROVED** [review]
NICE!
